### ISSUE TITLE
Epic: AI-First Manga Reader Features

### ISSUE THREAD

**Issue by HeartlessVeteran2** [Heartless-Veteran/Otaku-Reader#8]
---
name: ai-features
description: Comprehensive AI integration roadmap for Otaku-Reader using Gemini for cloud AI
labels: ["enhancement", "AI", "epic"]
---

# 🤖 AI Features Integration — Master Plan

**Epic:** AI-First Manga Reader Experience  
**Priority:** High  
**Estimated Effort:** 4-6 sprints  
**Target:** Personal use with local-first, privacy-focused AI  
**Cloud Provider:** Gemini (Google AI) — Developer Access ✅

---

## 🎯 Philosophy

Otaku-Reader is AI-first by design. Unlike public apps that limit AI to control costs, this personal reader maxes out AI capabilities using:
- **Local models** (Ollama) for privacy-sensitive operations
- **Gemini API** (Google AI) for heavy lifting — free tier available!
- **On-device ML** (ML Kit) for real-time features

Zero telemetry. Zero shared data. Your reading history stays yours.

---

### 1. Smart Recommendations Engine

**Local + Cloud Hybrid Architecture:**
```
Reading History → Local Filter (Ollama 3B) → Top 50 Candidates 
                                    ↓
                           Cloud Rank (Gemini 2.5 Pro/Flash)
                                    ↓
                           Final 10 Recommendations
```

**Capabilities:**
- [ ] **Taste Profile Learning**: Learns from your read/dropped/favorited manga
- [ ] **Because You Dropped X**: Suggests alternatives to dropped series
- [ ] **Completion Prediction**: "You'll finish this in 2 more sessions"
- [ ] **Mood-Based**: "Show me something lighthearted after that dark series"
- [ ] **Auto-Hide**: Hides manga you'll statistically never finish

**Tech Stack:**
- Local: Ollama with Llama 3.2 3B (embeddings + classification)
- Cloud: **Gemini API** (ranking + reasoning) — free tier: 60 requests/min
- Cache: Local SQLite with vector search (sqlite-vec)

---

### 2. Content Understanding Pipeline

**Auto-Analysis on Library Add:**

| Feature | Model | Location |
|---------|-------|----------|
| Genre/Tag Extraction | Fine-tuned classifier | Local (Ollama) |
| Synopsis Generation | Gemini 2.5 Flash | Cloud (free tier) |
| Art Style Classification | CLIP/ViT | Local |
| Character Identification | Gemini 2.5 Pro | Cloud (batch) |
| Mood/Tone Analysis | Sentiment classifier | Local |

**Database Schema:**
```sql
CREATE TABLE ai_analysis (
    manga_id INTEGER PRIMARY KEY,
    auto_tags TEXT,              -- AI-generated tags
    genres TEXT,                 -- Predicted genres
    mood TEXT,                   -- lighthearted, dark, thrilling, etc.
    art_style TEXT,              -- Shonen, Seinen, Josei, etc.
    summary TEXT,                -- AI-generated synopsis
    characters TEXT,             -- JSON: [{name, role, traits}]
    content_warnings TEXT,       -- AI-detected warnings
    quality_score REAL,          -- 0-1 predicted enjoyment
    analyzed_at INTEGER,
    analysis_version INTEGER
);
```

---

### 3. Reading Assistant

**In-Reader Features:**

#### SFX Translation (On-Device)
- [ ] **OCR**: ML Kit Text Recognition on manga panels
- [ ] **Translation**: Local LLM (Llama 3.2 1B) for Japanese → English
- [ ] **Overlay**: Non-intrusive translation bubbles
- [ ] **Cache**: Translated SFX stored locally per manga

#### Context Summaries
- [ ] **Previous Chapter Recap**: "Last time: MC fought the villain..."
- [ ] **Character Reminder**: "This is the sister who appeared in Ch. 5"
- [ ] **Plot Prediction**: "Based on tropes, expect a betrayal next"

**Implementation:**
```kotlin
// Summarize last 3 chapters before reading
suspend fun generateRecap(mangaId: Long): String {
    val recentChapters = getRecentChapters(mangaId, limit = 3)
    val ocrText = recentChapters.flatMap { extractText(it) }
    return ollama.summarize(ocrText, style = "manga_recap")
}
```

---

#### Auto-Categorization
- [ ] **Smart Collections**: "Unread action under 50 chapters"
- [ ] **Dynamic Categories**: Auto-created based on reading patterns
- [ ] **Seasonal Grouping**: "Summer 2024 reads"
- [ ] **Completion Likelihood**: Auto-sort by "will finish" score

#### Natural Language Library Search
```
"Find me isekai with OP MC but no harem, under 100 chapters"
"Show manga similar to Solo Leveling but completed"
"What was that manga with the time-traveling chef?"
```

**Implementation:**
- Local embeddings for all manga metadata
- Vector search with sqlite-vec
- NLP query parsing with local LLM

---

#### Predictive Pre-loading
- [ ] **Reading Pace Learning**: Downloads next N chapters based on your speed
- [ ] **WiFi Prediction**: "You're on WiFi, downloading next 5 chapters"
- [ ] **Battery Awareness**: Pauses downloads under 20%

#### Smart Cleanup
- [ ] **Access Pattern Analysis**: "You haven't touched this in 3 months"
- [ ] **Completion Prediction**: Deletes manga you're likely to drop
- [ ] **Storage Optimization**: Keeps favorites, removes stalled reads

---

### 6. Panel-by-Panel Reader (Advanced)

**ML-Based Panel Detection:**
- [ ] **Auto-Crop**: Detects panel boundaries
- [ ] **Tap-to-Advance**: Taps advance to next panel, not page
- [ ] **Webtoon Optimization**: Seamless vertical scrolling with panel hints
- [ ] **Double-Page Spread Handling**: AI detects spreads, shows full width

**Tech Stack:**
- TensorFlow Lite (on-device)
- Pre-trained manga panel detection model
- Custom training on your specific sources if needed

---

### Module Structure

```
ai/
├── local/                      # On-device AI
│   ├── embeddings/             # Text/image embeddings
│   ├── classification/         # Genre, mood, quality
│   ├── ocr/                    # SFX translation
│   └── panel/                  # Panel detection models
│
├── remote/                     # Cloud AI APIs
│   ├── google/                 # Gemini API (primary)
│   ├── openai/                 # OpenAI (optional fallback)
│   └── anthropic/              # Claude (optional)
│
├── pipeline/                   # Processing queues
│   ├── AnalysisWorker.kt       # Background manga analysis
│   ├── RecommendationEngine.kt # Periodic recs generation
│   └── SyncWorker.kt           # Cross-device AI state sync
│
└── cache/                      # Local AI storage
    ├── vector/                 # sqlite-vec embeddings
    └── models/                 # Downloaded TFLite models
```

### Data Flow

```
1. Manga Added to Library
         ↓
2. Local Analysis (immediate)
   - Extract text from first chapter
   - Classify genre/mood
   - Generate embeddings
         ↓
3. Queue for Cloud Analysis (background)
   - Generate synopsis (if missing) via Gemini
   - Identify characters via Gemini Pro
   - Quality scoring
         ↓
4. Store Results
   - SQLite: ai_analysis table
   - Vector DB: embeddings for search
         ↓
5. Update Recommendations
   - Include in taste profile
   - Generate new suggestions
```

---

### Phase 1: Foundation (Sprint 1-2)
- [ ] Set up Ollama on home server/Pi
- [ ] Set up Gemini API key (free tier)
- [ ] Create `ai` module structure
- [ ] Implement local embeddings (sqlite-vec)
- [ ] Basic manga metadata analysis pipeline

### Phase 2: Recommendations (Sprint 3-4)
- [ ] Taste profile learning
- [ ] Local recommendation engine
- [ ] Gemini ranking integration
- [ ] Library UI for recommendations

### Phase 3: Content Understanding (Sprint 5-6)
- [ ] Auto-tagging system
- [ ] Synopsis generation via Gemini Flash
- [ ] Mood classification
- [ ] Art style detection

### Phase 4: Reader AI (Sprint 7-8)
- [ ] SFX translation with ML Kit OCR
- [ ] Context summaries via Gemini
- [ ] Chapter recaps
- [ ] Reading pace learning

### Phase 5: Advanced Features (Sprint 9-10)
- [ ] Panel detection
- [ ] Natural language search
- [ ] Smart downloads
- [ ] Cross-device AI sync

---

## 🛠️ Tech Stack

| Component | Technology | Cost |
|-----------|------------|------|
| Local LLM | Ollama + Llama 3.2 | $0 (self-hosted) |
| On-device OCR | ML Kit | $0 |
| Vector Search | sqlite-vec | $0 |
| **Cloud LLM** | **Gemini API** | **$0 (developer access: higher limits)** |
| Cloud LLM (paid) | Gemini 2.5 Pro | ~$0-5/month (light usage) |
| Panel Detection | TensorFlow Lite | $0 |
| Image Analysis | CLIP (local) | $0 |

### Gemini Pricing (Developer Access ✅)

| Model | Developer Access | Rate Limits |
|-------|-----------------|-------------|
| Gemini 2.5 Flash | ✅ Included | Higher than free tier |
| Gemini 2.5 Pro | ✅ Included | Higher than free tier |
| Embeddings | ✅ Included | Increased quota |
| Experimental Models | ✅ Early access | Try latest features |

**Benefits of Developer Access:**
- Higher rate limits than standard free tier
- Access to experimental/preview models
- Priority API access
- Early access to new features

**For personal manga reader: Developer access = no rate limit concerns.**

---

## 🔒 Privacy Considerations

1. **No Cloud Required**: All core features work offline
2. **Opt-in Cloud**: Advanced features use your personal Gemini API key
3. **Local-First**: Data never leaves your devices unless you explicitly call Gemini
4. **Encrypted Sync**: AI state syncs via Syncthing/Tailscale (P2P, no cloud)
5. **No Telemetry**: Zero analytics, zero tracking
6. **Gemini Data Policy**: Google doesn't train on API data (https://ai.google.dev/gemini-api/terms)

---

## 🎯 Success Metrics

- [ ] Recommendations: 80%+ click-through rate
- [ ] Auto-tags: 90%+ accuracy vs manual tagging
- [ ] SFX Translation: <2s per panel, 85%+ accuracy
- [ ] Panel Detection: 95%+ accuracy on standard manga
- [ ] Library Search: <500ms response time
- [ ] Gemini API usage: Unlimited (developer access)

---

## 📚 Related Issues

- #5 — Initial repository setup (completed)
- Future: Cross-device sync architecture
- Future: Source extension API with AI hooks

---

### Why Gemini?
- ✅ **Developer Access**: Higher rate limits, experimental models
- ✅ **Free tier**: 60 requests/min base (sufficient for personal use)
- ✅ **Speed**: Faster than OpenAI for similar quality
- ✅ **Integration**: Native Android SDK available
- ✅ **Multimodal**: Great for image (manga) + text analysis
- ✅ **No training on API data**: Privacy preserved

### Developer Access Benefits
| Feature | Standard | Developer Access |
|---------|----------|------------------|
| Rate limits | 60 req/min | **Higher limits** |
| Experimental models | ❌ | ✅ Early access |
| Priority access | ❌ | ✅ Priority |
| New features | Standard rollout | **Early access** |

### Alternative Providers (Optional)
| Provider | Use Case | Cost |
|----------|----------|------|
| OpenAI GPT-4o | Fallback if Gemini down | $10-20/month |
| Anthropic Claude | Long context analysis | $10-20/month |
| Ollama (local) | 100% offline operation | $0 |

**Default: Gemini (free tier)**  
**Fallback: Ollama local models**

---

*Created by Kimi Claw — Otaku-Reader AI Architect* 🤖❤️‍🔥  
*Updated: Using Gemini for cloud AI — free tier sufficient for personal use*
